esx_service/vmdk_ops.py

```diff
@@ -691,6 +710,38 @@ def getStatusAttached(vmdk_path):
         uuid = None
     return attached, uuid, attach_as
 
+def handle_stale_attach(vmdk_path, kv_uuid):
+       '''
+       Clear volume state for cases where the VM that attached the disk
+       earlier is powered off or removed, detach the disk from the VM
+       if its powered off.
+       '''
+       cur_vm = findVmByUuid(kv_uuid)
+
+       if cur_vm:
+          # Detach the disk only if VM is powered off
+          if cur_vm.runtime.powerState == VM_POWERED_OFF:
+             logging.info("Detaching disk %s from VM(powered off) - %s\n",
+                             vmdk_path, cur_vm.config.name)
+             device = findDeviceByPath(vmdk_path, cur_vm)
+             if device:
+                msg = disk_detach(vmdk_path, cur_vm, device)
+                if msg:
+                   msg += " failed to detach disk {0} from VM={1}.".format(vmdk_path,
+                                                                           cur_vm.config.name)
+                   return err(msg)
+             else:
+                logging.warning("Failed to find disk %s in powered off VM - %s, resetting volume metadata\n",
+                                vmdk_path, cur_vm.config.name)
+                reset_vol_meta(vmdk_path)
+          else:
+             msg = "Disk {0} already attached to VM={1}".format(vmdk_path,
+                                                                cur_vm.config.name)
+             return err(msg)
+       else:
+          logging.warning("Failed to find VM %s that attached the disk %s, resetting volume metadata",
+                          cur_vm.config.name, vmdk_path)
+          reset_vol_meta(vmdk_path)
```

Comment:
Cross host clean up is a bigger issue for the plugin and needs plugin level communication across nodes. All of these managed objects we use to config the VM are per node only. And only if the VM is found registered on the node. Otherwise the VM is simply not found. The proposed change will not mount if the VM thats holding the disk isn't found on the local host.